### DISCUSSION TITLE
Add context (and request?) to ErrorBoundary

### DISCUSSION THREAD

**Discussion by jamesarosen** [remix-run/remix#6297] [Proposals]
## Problem

`ErrorBoundary` has the following type:

```ts
type RemixErrorBoundaryProps = React.PropsWithChildren<{
    location: Location;
    component: ErrorBoundaryComponent;
    error?: Error;
}>;

type RemixErrorBoundaryState = {
    error: null | Error;
    location: Location;
};

export declare class RemixErrorBoundary extends React.Component<RemixErrorBoundaryProps, RemixErrorBoundaryState> {
    constructor(props: RemixErrorBoundaryProps);
    static getDerivedStateFromError(error: Error): {
        error: Error;
    };
    static getDerivedStateFromProps(props: RemixErrorBoundaryProps, state: RemixErrorBoundaryState): {
        error: Error | null;
        location: Location;
    };
    render(): string | number | boolean | React.ReactFragment | JSX.Element | null | undefined;
}
```

Since the props don't include `context`, any data or services that require `context` are unavailable in error boundaries.

In many environments (including Cloudflare Pages), `process.env` doesn't exist; environment variables are instead attached to the `context`.

This makes it very difficult to use an error boundary to report errors to third-party trackers like Sentry, Rollbar, or LogRocket.

### Proposal

Add `context: AppLoadContext` to the `ErrorBoundary` props. Better yet, if it's not prohibitive, add all of `DataFunctionArgs` (`{ request, context, params }`).

**Comment by jamesarosen** [comment]
One alternative is to include the necessary `ENV` vars in the thrown `error`:

```ts
export function loader({ context }) {
  try {
    doSomethingDangerous()
  } catch (e) {
    throw new Error(JSON.stringify({ message: e.message, env: context.ENV }))
  }
}

export function ErrorBoundary({ error }: { error: Error }) {
  let env, message
  try {
    const parsed = JSON.parse(error.message)
    env = parsed.env
    message = parsed.message
  } catch {
    env = {}
    message = error.message
  }

  doSomethingWith(env)

  return <html>{message}</html>
}
```

This requires wrapping every `action`, `loader`, and `meta` throughout the app, which is a bit annoying.

**Comment by sergiodxa** [comment]
ErrorBoundary runs both server and client side, so you can't pass the AppLoadContext because that's a server-only thing.

Doing this will limit error boundaries to only work server-side and never hydrate your app.

Another way this may be able to work is once RSC are stable and Remix supports them, error boundaries could be considered server components and have access to every AppLoadContext.

**Reply by jamesarosen** [reply]
> ErrorBoundary runs both server and client side, so you can't pass the AppLoadContext because that's a server-only thing

Great point.

An alternative solution would be loader data for an ErrorBoundary:

```tsx
// app/routes/my-route.tsx

export function errorLoaderData({ request, context, params }) {
  return { ENV: context.env }
}

export function ErrorBoundary() {
  const { ENV } = useErrorLoaderData()
  ...
}
```

Another alternative would be to pass the `AppLoadContext` to `entry.server.tsx`:

```tsx
// app/entry.server.tsx

export default function handleRequest(
  request: Request,
  responseStatusCode: number,
  responseHeaders: Headers,
  remixContext: EntryContext,
  context: AppLoadContext
) {
  const markup = renderToString(
    <EnvironmentVariablesContext env={context.env}>
      <RemixServer context={remixContext} url={request.url} />,
    <EnvironmentVariablesContext>
  )
  return new Response('<!DOCTYPE html>' + markup, {
    status: responseStatusCode,
    headers: responseHeaders,
  })
}
```

**Reply by sergiodxa** [reply]
> An alternative solution would be loader data for an ErrorBoundary 

You can't access loader data in ErrorBoundary, because the trigger of the error boundary could be the loader failing.

```ts
export async function loader() {
  throw new Error("Something failed in the loader")
}

export function ErrorBoundary() {
  // this was rendered because loader failed, so there's no loader data
}
```

What you could do is to access a parent route loader data, so if you have `root` and `routes/_index` and the `routes/_index` loader fails and renders an ErrorBoundary defined in that route (not in root) you will be able to access `root`'s loader data, if the `root` files the ErrorBoundary of root can't access any loader data.

> Another alternative would be to pass the AppLoadContext to entry.server.tsx:

This is supported on the latest version of Remix

**Reply by jamesarosen** [reply]
> You can't access loader data in ErrorBoundary

This is why I had suggested the new `errorLoaderData`.

> What you could do is to access a parent route loader data, so if you have root and routes/_index and the routes/_index

True. The downside is that I need to split a route that has `ENV` and business data into two routes: `root` and `routes/_myPathlessGlobalLayoutRoute`. This adds to the routing hierarchy. https://github.com/remix-run/remix/discussions/5431 could mitigate this.

**Reply by jamesarosen** [reply]
> [passing the `AppLoadContext` to `entry.server.tsx`] is supported on the latest version of Remix

This is excellent news! https://github.com/remix-run/remix/commit/98bdd4c78e6b0f76273ed813c18d73609b3091ea has the commit and it does exactly what I need!
